lib/logstash/plugin_mixins/rabbitmq_connection.rb

```diff
@@ -0,0 +1,149 @@
+
+      end
+
+      def conn_str
+        "amqp://#{@user}@#{@host}:#{@port}#{@vhost}"
+      end
+
+      def close_connection
+        @rabbitmq_connection_stopping = true
+        @hare_info.channel.close if channel_open?
+        @hare_info.connection.close if connection_open?
+      end
+
+      def rabbitmq_settings
+        return @rabbitmq_settings if @rabbitmq_settings
+
+        s = {
+          :vhost => @vhost,
+          :host  => @host,
+          :port  => @port,
+          :user  => @user,
+          :automatic_recovery => @automatic_recovery,
+          :pass => @password ? @password.value : "guest",
+        }
+        s[:tls] = @ssl if @ssl
+        @rabbitmq_settings = s
+      end
+
+      def connect!
+        @hare_info = connect() unless @hare_info # Don't duplicate the conn!
+      rescue MarchHare::Exception => e
+        @logger.error("RabbitMQ connection error, will retry.",
+                      :message => e.message,
+                      :exception => e.class.name,
+                      :backtrace => e.backtrace)
+
+        sleep_for_retry
+        retry
+      end
+
+      def channel_open?
+        @hare_info && @hare_info.channel && @hare_info.channel.open?
+      end
+
+      def connection_open?
+        @hare_info && @hare_info.connection && @hare_info.connection.open?
+      end
+
+      def connected?
+        return nil unless @hare_info && @hare_info.connection
+        @hare_info.connection.connected?
+      end
+
+      private
+
+      def declare_exchange!(channel, exchange, exchange_type, durable)
+        @logger.debug("Declaring an exchange", :name => exchange,
+                      :type => exchange_type, :durable => durable)
+        exchange = channel.exchange(exchange, :type => exchange_type.to_sym, :durable => durable)
+        @logger.debug("Exchange declared")
+        exchange
+      rescue StandardError => e
+        @logger.error("Could not declare exchange!",
+                      :exchange => exchange, :type => exchange_type,
+                      :durable => durable, :error_class => e.class.name,
+                      :error_message => e.message, :backtrace => e.backtrace)
+        raise e
+      end
+
+      def connect
+        @logger.debug? && @logger.debug("Connecting to RabbitMQ. Settings: #{rabbitmq_settings.inspect}")
+
+        connection = MarchHare.connect(rabbitmq_settings)
+        connection.on_blocked { @logger.warn("RabbitMQ output blocked! Check your RabbitMQ instance!") }
+        connection.on_unblocked { @logger.warn("RabbitMQ output unblocked!") }
+
+        channel = connection.create_channel
+        @logger.info("Connected to RabbitMQ at #{rabbitmq_settings[:host]}")
+
+        HareInfo.new(connection, channel)
+      end
+
+      def sleep_for_retry
+        return if @connect_retry_interval < 1
+
+        @connect_retry_interval.times do
+          break if @rabbitmq_connection_stopping
+          sleep 1
+        end
```

Comment:
`stud` already implement this logic.

`Stud.stoppable_sleep(10) { @rabbitmq_connection_stopping }`